core/coreunix/add.go

```diff
@@ -90,49 +150,176 @@ func AddWrapped(n *core.IpfsNode, r io.Reader, filename string) (string, *merkle
 	return gopath.Join(k.String(), filename), dagnode, nil
 }
 
-func add(n *core.IpfsNode, reader io.Reader) (*merkledag.Node, error) {
-	mp := n.Pinning.GetManual()
+func NewAdder(ctx context.Context, n *core.IpfsNode, out chan interface{}, progress bool, hidden bool, trickle bool, wrap bool, chunker string) Adder {
+	e := dagutils.NewDagEditor(newMemoryDagService(), newDirNode())
+	return &adder{
+		ctx:      ctx,
+		node:     n,
+		editor:   e,
+		out:      out,
+		progress: progress,
+		hidden:   hidden,
+		trickle:  trickle,
+		wrap:     wrap,
+		chunker:  chunker,
+	}
+}
+
+// Internal structure for holding the switches passed to the `add` call
+type adder struct {
+	ctx      context.Context
+	node     *core.IpfsNode
+	editor   *dagutils.Editor
+	out      chan interface{}
+	progress bool
+	hidden   bool
+	trickle  bool
+	wrap     bool
+	chunker  string
+}
+
+// Perform the actual add & pin locally, outputting results to reader
+func (params *adder) add(reader io.Reader, pinDirect bool) (*merkledag.Node, error) {
+	mp := params.node.Pinning.GetManual()
 
+	chnk, err := chunk.FromString(reader, params.chunker)
+	if err != nil {
+		return nil, err
+	}
+
+	cb := importer.PinIndirectCB(mp)
+	if pinDirect {
+		cb = importer.BasicPinnerCB(mp)
+	}
+
+	if params.trickle {
+		return importer.BuildTrickleDagFromReader(
+			params.node.DAG,
+			chnk,
+			cb,
+		)
+	}
 	return importer.BuildDagFromReader(
-		n.DAG,
-		chunk.DefaultSplitter(reader),
-		importer.PinIndirectCB(mp),
+		params.node.DAG,
+		chnk,
+		cb,
 	)
 }
 
-func addNode(n *core.IpfsNode, node *merkledag.Node) error {
-	if err := n.DAG.AddRecursive(node); err != nil { // add the file to the graph + local storage
+func (params *adder) PinRoot() error {
+	root := params.editor.GetNode()
+
+	// if not wrapping, AND one root file, use that hash as root.
+	if !params.wrap && len(root.Links) == 1 {
+		var err error
+		root, err = root.Links[0].GetNode(params.ctx, params.editor.GetDagService())
+		// no need to output, as we've already done so.
+		if err != nil {
+			return err
+		}
+	} else {
+		// otherwise need to output, as we have not.
+		if err := outputDagnode(params.out, "", root); err != nil {
+			return err
+		}
+	}
```

Comment:
can we keep this in a `RootNode()` function? this isn't specific to pinning, and it makes it possible to grab the root without pinning (i.e. when the option is false).

```go
func (params *adder) RootNode() (*dag.Node, error) {
	// for memoizing
	if params.root != nil {
		return params.root, nil
	}

	root := params.editor.GetNode()  

	// if not wrapping, AND one root file, use that hash as root.
	if !params.wrap && len(root.Links) == 1 {
		var err error
		root, err = root.Links[0].GetNode(params.ctx, params.editor.GetDagService())
		// no need to output, as we've already done so.
		params.root = root
		return root, err
	}

	// otherwise need to output, as we have not.
	err := outputDagnode(params.out, "", root)
	params.root = root
	return root, err
}
```

and call `root := params.RootNode()` in `PinRoot()` or something